### ISSUE TITLE
Regression introduced by new min max validators

### ISSUE THREAD

**Issue by PowerKiKi** [angular/angular#42267]
# Bug Report

### Affected Package

@angular/forms 12+

### Is this a regression?

Yes, the previous version in which this bug was not present was: 11

### Description
@angular/forms 12 introduced min and max validators based on the min and max attributes presence in #39063.

Before that, the browser would react to min/max attributes by doing validation and also by preventing to introduce invalid value. Both of those **native features are now impossible to leverage**.

Also, before that, it was possible to deactivate the browser native features with a null value, such as `[min]="null"` or `[min]="undefined"`. This is now impossible to do in strict mode compilation, because the new directive only accept `string | number` and would **result in a compilation error**. Possible temporary workaround for that is to use `$any` casting similar to `[min]="configuration.min ?? $any(undefined)"`

![image](https://user-images.githubusercontent.com/72603/119329150-def05580-bcbf-11eb-8131-d490496b6b82.png)

### Suggested fixes

I would suggest that both directives create, and maintain, the equivalent attribute on the HTMLInputElement, so that browsers are able to react accordingly and adapt their UX.

And I would suggest to loosen the input type like so:

```diff
- @Input() min!: string|number;
+ @Input() min: string|number|null|undefined;
```

If those suggestions sound mergeable I could create PRs for it.

## Minimal Reproduction

- https://stackblitz.com/edit/min-max-regression-angular-11
- https://stackblitz.com/edit/min-max-regression-angular-12, same as above, but with strict mode disabled, and Angular 12

## Exception or Error

```
Type 'number | null' is not assignable to type 'string | number'.
```

## Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 12.0.1
Node: 14.17.0
Package Manager: yarn 1.22.5
OS: linux x64

Angular: 12.0.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, localize, material
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1200.1
@angular-devkit/build-angular   12.0.1
@angular-devkit/core            12.0.1
@angular-devkit/schematics      12.0.1
@angular/flex-layout            12.0.0-beta.34
@schematics/angular             12.0.1
ng-packagr                      12.0.0
rxjs                            6.6.7
typescript                      4.2.4

</code></pre>

**Comment by PowerKiKi** [comment]
@sonukapoor, since you wrote the PR that brought this regression, would you have any particular opinions about this issue ? were those use-cases considered in your PR ?

**Comment by iRealNirmal** [comment]
@petebacondarwin I can look into this.

**Comment by sonukapoor** [comment]
> @sonukapoor, since you wrote the PR that brought this regression, would you have any particular opinions about this issue ? were those use-cases considered in your PR ?

@PowerKiKi I do not think this use case of null/undefined was considered. 
@iRealNirmal let us know if you need any help with that.

**Comment by iRealNirmal** [comment]
> > @sonukapoor, since you wrote the PR that brought this regression, would you have any particular opinions about this issue ? were those use-cases considered in your PR ?
> 
> @PowerKiKi I do not think this use case of null/undefined was considered.
> @iRealNirmal let us know if you need any help with that.

Sure @sonukapoor

**Comment by AndrewKushnir** [comment]
Hi, just want to provide a bit more context.

The `min`/`max` validators were modeled after `minlength`/`maxlength` ones, which have the input type defined as `string|number`:

https://github.com/angular/angular/blob/4ac55ca676769a54a76c5881d40109b556f497b7/packages/forms/src/directives/validators.ts#L553-L554

However in the code below the input value is checked against `null` and `undefined` values as well:

https://github.com/angular/angular/blob/4ac55ca676769a54a76c5881d40109b556f497b7/packages/forms/src/directives/validators.ts#L569-L571

So the `minlength` and `maxlength` directives would have the same problem from types perspective, but would work as expected (no validation happening) if the value is `null` or `undefined`.

It looks like the `min` and `max` validators work with `null` and `undefined` by accident where these values are [converted to `NaN`](https://github.com/angular/angular/blob/master/packages/forms/src/directives/validators.ts#L241) and after that the comparison with a number always gives `false`, so the underlying [minValidator](https://github.com/angular/angular/blob/bf505ad7071f54f44b4bfc18155e55599002e668/packages/forms/src/validators.ts#L424) and [maxValidator](https://github.com/angular/angular/blob/bf505ad7071f54f44b4bfc18155e55599002e668/packages/forms/src/validators.ts#L440) functions return "validation is successful" result.

Looking at other validator directives in the [packages/forms/src/directives/validators.ts](https://github.com/angular/angular/blob/master/packages/forms/src/directives/validators.ts) file, some of them have special checks for when validator should be enabled, such as the [required](https://github.com/angular/angular/blob/master/packages/forms/src/directives/validators.ts#L357) validator, but in some cases (for example in [the pattern validator](https://github.com/angular/angular/blob/master/packages/forms/src/directives/validators.ts#L727)), the validation will always happen and there is no way to conditionally enable/disable it.

My conclusion is that we should revisit all existing form validators from that perspective and make sure that the logic is consistent. Specifically verify (by adding tests) that there is a way to conditionally disable validation, update input types accordingly (to allow for `null|undefined` types) and mention this behavior in docs.

**Comment by Airblader** [comment]
I would think you can workaround this issue by binding to `[attr.min]` instead, fwiw.

**Comment by PowerKiKi** [comment]
> `min` and `max` validators work with `null` and `undefined` by accident

Exactly my conclusion too. I read the code first and expected it to fail, but was surprised when I tried and it actually worked. That's definitely something that will need to be supported explicitly, and not accidentally.

> I would think you can workaround this issue by binding to `[attr.min]` instead, fwiw.

Excellent suggestion ! I wish I thought about that yesterday. This allows to keep using the browsers native features (:+1:). And it also avoid having a duplicated validators (:+1:): the one I already explicitly added by code before Angular 12, and the one auto-magically injected since Angular 12. Thanks @Airblader !

I would go as far as saying that this is not just a workaround, but it actually **is the proper permanent solution** for people who manage their validator by code (as opposed to templates). And it should probably have been mentioned in the migration notes.

And I'll go a bit off-topic, but this reminds me of the situation of the `required` validator, https://github.com/angular/components/issues/2574, where the validator should impact the template (to show the `*`), and not the other way round. Here too, for people working with reactive forms, the validator defined by code should drive the template.

As far as I am concerned, I am no longer impacted by this issue, but I still think it should be fixed.

**Comment by iRealNirmal** [comment]
Hi @PowerKiKi thanks for reporting the issues. As I can see you faced multiple issue with example you provided but I have question  can you help me with it

1. Issue regarding compilation error in strict mode which I was able to reproduce as well in stack blitz but when I am doing build in local it's working perfectly fine.

@sonukapoor @AndrewKushnir do you also have any input for this behaviour ? I am not sure but can this be related to stackblitz ?

2. Issue min arrow not working in case minimum value is 0

3. Issue 3  regarding invalid not being set when  value is 0.

I will work on fixing issue 2 and 3 while you provide more input for issue 1.

**Comment by AndrewKushnir** [comment]
> Issue regarding compilation error in strict mode which I was able to reproduce as well in stack blitz but when I am doing build in local it's working perfectly fine.

@iRealNirmal can you share the error message that you see on stackblitz and which code (in the app itself) triggers it?

**Comment by iRealNirmal** [comment]
yes @AndrewKushnir here is stackblitz which isn't working. It's coming due to same thing as mentioned by @PowerKiKi. But as I told, this same example if I download and try to run  in local system and it's working.

[https://stackblitz.com/edit/min-max-regression-angular-12-srr4vy?file=src%2Fapp%2Fapp.component.ts](https://stackblitz.com/edit/min-max-regression-angular-12-srr4vy?file=src%2Fapp%2Fapp.component.ts)

Even though stackblitz is failing from my perspective it's working as expected because it's fulfilling purpose of strict type.

**Comment by iRealNirmal** [comment]
@AndrewKushnir For issue 2 and 3  what I observed `ɵɵattribute` from core package is getting value null and it's removing min attribute and it will  remove all other attributes as well.

I am trying to resolve it but I have to still figure out where it's getting  set null. @AndrewKushnir If you have any other input as well regarding this method getting  invoked then it will help me to find a fix  for this issue.

**Comment by iRealNirmal** [comment]
thanks @PowerKiKi  this helped.

**Comment by petebacondarwin** [comment]
As discussed offline the `min` attribute is being removed when the value is 0 is because of this `host` binding:

```ts
@Directive({
  selector:
      'input[type=number][min][formControlName],input[type=number][min][formControl],input[type=number][min][ngModel]',
  providers: [MIN_VALIDATOR],
  host: {'[attr.min]': 'min ? min : null'}
})
```

**Comment by iRealNirmal** [comment]
yes thanks @petebacondarwin for it.

**Comment by PowerKiKi** [comment]
This issue is **not** solved until #42378 is also merged. But I cannot re-open the issue myself.
